core/src/main/scala/cats/collections/HashSet.scala

```diff
@@ -0,0 +1,1144 @@
+
+    private var currentDepth: Int = -1
+
+    private val nodeStack: Array[Node[A]] =
+      new Array(Node.MaxDepth)
+
+    private val nodeIndicesAndLengths: Array[Int] =
+      new Array(Node.MaxDepth * 2)
+
+    def this(rootNode: Node[A]) = {
+      this()
+      if (rootNode.hasNodes) pushNode(rootNode)
+      if (rootNode.hasValues) pushValues(rootNode)
+    }
+
+    final private def pushNode(node: Node[A]): Unit = {
+      currentDepth += 1
+
+      val cursorIndex = currentDepth * 2
+      val lengthIndex = currentDepth * 2 + 1
+
+      nodeStack(currentDepth) = node
+
+      nodeIndicesAndLengths(cursorIndex) = 0
+      nodeIndicesAndLengths(lengthIndex) = node.nodeCount
+    }
+
+    final private def pushValues(node: Node[A]): Unit = {
+      currentNode = node
+      currentValuesIndex = 0
+      currentValuesLength = node.valueCount
+    }
+
+    final private def getMoreValues(): Boolean = {
+      var foundMoreValues = false
+
+      while (!foundMoreValues && currentDepth >= 0) {
+        val cursorIndex = currentDepth * 2
+        val lengthIndex = currentDepth * 2 + 1
+
+        val nodeIndex = nodeIndicesAndLengths(cursorIndex)
+        val nodeLength = nodeIndicesAndLengths(lengthIndex)
+
+        if (nodeIndex < nodeLength) {
+          val nextNode = nodeStack(currentDepth)
+            .getNode(nodeIndex)
+
+          if (nextNode.hasNodes) {
+            pushNode(nextNode)
+          }
+
+          if (nextNode.hasValues) {
+            pushValues(nextNode)
+            foundMoreValues = true
+          }
+
+          nodeIndicesAndLengths(cursorIndex) += 1
+
+        } else {
+          currentDepth -= 1
+        }
+      }
+
+      foundMoreValues
+    }
+
+    final override def hasNext: Boolean =
+      (currentValuesIndex < currentValuesLength) || getMoreValues()
+
+    final override def next(): A = {
+      if (!hasNext) throw new NoSuchElementException
+      val value = currentNode.getValue(currentValuesIndex)
+      currentValuesIndex += 1
+      value
+    }
+  }
+
+  implicit val catsDataUnorderedFoldableForHashSet: UnorderedFoldable[HashSet] =
+    new UnorderedFoldable[HashSet] {
+      override def isEmpty[A](fa: HashSet[A]): Boolean = fa.isEmpty
+
+      override def nonEmpty[A](fa: HashSet[A]): Boolean = fa.nonEmpty
+
+      override def size[A](fa: HashSet[A]): Long = fa.size.toLong
+
+      override def exists[A](fa: HashSet[A])(p: A => Boolean): Boolean = fa.iterator.exists(p)
+
+      override def forall[A](fa: HashSet[A])(p: A => Boolean): Boolean = fa.iterator.forall(p)
+
+      override def count[A](fa: HashSet[A])(p: A => Boolean): Long =
+        fa.iterator.foldLeft(0L) { case (c, a) => if (p(a)) c + 1L else c }
+
+      def unorderedFoldMap[B, C](fa: HashSet[B])(f: B => C)(implicit C: CommutativeMonoid[C]): C =
+        C.combineAll(fa.iterator.map(f))
+    }
+
+  implicit def catsDataCommutativeMonoidForHashSet[A](implicit hash: Hash[A]): CommutativeMonoid[HashSet[A]] =
```

Comment:
there is a question here as to which is the "correct" default monoid.

You can think of the set as an infinite vector with all entries that are present having the boolean value true, otherwise false (the contains function). You could OR those booleans (what this monoid is doing), but we could also AND them (intersection) or XOR them (exclusive or).

I tend to agree union is the most common, but we could also imagine having a non-implicit intersectionMonoid  which is also a `CommutativeMonoid`.

Lastly, you can implement a DistributiveLattice: https://github.com/typelevel/cats/blob/main/algebra-core/src/main/scala/algebra/lattice/DistributiveLattice.scala where join is union and meet is intersection.

Reply:
Yes that's worth exploring - I'm not sure now why I decided to implement `CommutativeMonoid` specifically here

Reply:
I've added a `DistributiveLattice` instance in https://github.com/typelevel/cats-collections/pull/533/commits/aaa8a0a07ba2cbe98e436a50ea11f83a42d7853b. The default `CommutativeMonoid` instance is still the `HashSetUnionMonoid` but users can instantiate the `HashSetIntersectionMonoid` if they need it or use the `DistributiveLattice`.